File: nodes/js/rope-node.js
// Rope Gateway
const ROPELIVE = 'https://secure.rope.live'

// Check for browser
var window = window || {}

// Get kite.js
const Kite = window.Kite || require('kite.js').Kite
const inBrowser = !!window.Kite

// Create dummy process
if (inBrowser) {
  process = { env: { ROPEHOST: ROPELIVE, ROPEDEBUG: Kite.DebugLevel.INFO } }
  document.body.appendChild(
    document.createTextNode('Rope-Node-JS: Check console logs for details.')
  )
}

// Defaults for Kite instance
const NAME = 'rope-node-js'
const ROPEHOST = process.env.ROPEHOST || ROPELIVE
const LOG_LEVEL = process.env.ROPEDEBUG || Kite.DebugLevel.INFO
const ENVIRONMENT = inBrowser ? 'Browser' : 'Node.js ' + process.version
const AUTO_RECONNECT = true

// The rope-node api
const api = {
  'rope.identified': res => {
    if (res.environment) {
      kite.environment = res.environment
    }
    kite.emit('info', `Identified as ${res.id} now!`)
  },
  'rope.identify': (id, callback) => {
    kite.emit('info', 'Identify requested!')
    const kiteInfo = {
      api: Object.keys(api),
      kiteInfo: kite.getKiteInfo(),
      signatures: {
        square: 'Number, Function',
      },
    }
    if (inBrowser) kiteInfo.useragent = navigator.userAgent
    callback(null, kiteInfo)
  },
  'rope.notify': notification => {
    console.log('Notification:', notification)
  },
  square: (number, callback) => {
    callback(null, number * number)
  },
}

// Create the rope-node Kite instance
var kite = new Kite({
  url: ROPEHOST,
  transportClass: Kite.transport.SockJS,
  name: NAME,
  logLevel: LOG_LEVEL,
  environment: ENVIRONMENT,
  autoReconnect: AUTO_RECONNECT,
  api: api,
})

// Connect!
kite.connect()

File: nodes/go/rope-node.go
package main

import (
	"github.com/koding/kite"
	"os"
	"runtime"
)

func main() {
	r := kite.New("dope", "0.0.0")
	r.Config.Environment = runtime.Version()

	kiteURL := os.Getenv("ROPEHOST")
	if kiteURL == "" {
		kiteURL = "https://secure.rope.live"
	}
	l := r.NewClient(kiteURL)
	l.Reconnect = true

	api := map[string]kite.HandlerFunc{
		"square": func(req *kite.Request) (interface{}, error) {
			number := req.Args.One().MustFloat64()
			result := number * number
			return result, nil
		},
		"identified": func(req *kite.Request) (interface{}, error) {
			var args struct {
				Id string `json:"id"`
			}
			req.Args.One().MustUnmarshal(&args)
			r.Log.Info("Identified as %v now!", args.Id)
			return nil, nil
		},
	}

	for method, f := range api {
		r.HandleFunc(method, f)
	}

	r.HandleFunc("identify", func(req *kite.Request) (interface{}, error) {
		r.Log.Info("Identify requested!")
		funcs := make([]string, 0, len(api))
		for method := range api {
			funcs = append(funcs, method)
		}
		return map[string]interface{}{
			"kiteInfo": r.Kite(),
			"api":      funcs,
		}, nil
	})

	connection, err := l.DialForever()
	if err != nil {
		r.Log.Fatal(err.Error())
	}
	<-connection

	r.Run()
}
